src/utils/use-onboarding-navigation-flow.ts

```diff
@@ -0,0 +1,196 @@
+import {useGeolocationState} from '@atb/GeolocationContext';
+import {useRemoteConfig} from '@atb/RemoteConfigContext';
+
+import {useAuthState} from '@atb/auth';
+import {useShouldShowShareTravelHabitsScreen} from '@atb/beacons/use-maybe-show-share-travel-habits-screen';
+import {
+  useNotifications,
+  usePushNotificationsEnabled,
+} from '@atb/notifications';
+import {RootNavigationProps, RootStackParamList} from '@atb/stacks-hierarchy';
+
+import {
+  useTicketingState,
+  filterValidRightNowFareContract,
+} from '@atb/ticketing';
+import {useTimeContextState} from '@atb/time';
+import {useNavigation, StackActions} from '@react-navigation/native';
+
+import {useCallback, useEffect, useState} from 'react';
+import {InteractionManager} from 'react-native';
+
+type ScreenProps =
+  | {
+      screenName?: keyof RootStackParamList;
+      params?: any;
+    }
+  | undefined;
+export type GoToScreenType = (screenName: any, params?: any) => void;
+
+// note: useThisHookInstanceForSessionCounting should only be true for one instance
+export const useOnboardingNavigationFlow = (
+  useThisHookInstanceForSessionCounting = false,
+) => {
+  const navigation = useNavigation<RootNavigationProps>();
+  const {enable_extended_onboarding} = useRemoteConfig();
+  const {onboarded: loginOnboardingCompleted} = useAppState();
+
+  const shouldShowTravelTokenOnboarding = useShouldShowTravelTokenOnboarding();
+
+  const shouldShowLocationOnboarding = useShouldShowLocationOnboarding();
+
+  const shouldShowShareTravelHabitsScreen =
+    useShouldShowShareTravelHabitsScreen(useThisHookInstanceForSessionCounting);
+
+  const shouldShowNotificationPermissionScreen =
+    useShouldShowNotificationPermissionScreen();
+
+  const getNextOnboardingScreen = useCallback(
+    (comingFromScreenName?: keyof RootStackParamList): ScreenProps => {
+      let screenName: keyof RootStackParamList | undefined = undefined;
+      let params = undefined;
+      if (!loginOnboardingCompleted) {
+        if (enable_extended_onboarding) {
+          screenName = 'Root_OnboardingStack';
+        } else {
+          screenName = 'Root_LoginOptionsScreen';
+          params = {};
+        }
+      } else {
+        const orderedOnboardingScreensAfterLogin: {
+          shouldShow: boolean;
+          screenName: keyof RootStackParamList;
+          params?: any;
+        }[] = [
+          {
+            shouldShow: shouldShowTravelTokenOnboarding,
+            screenName: 'Root_ConsiderTravelTokenChangeScreen',
+          },
+          {
+            shouldShow: shouldShowLocationOnboarding,
+            screenName: 'Root_LocationWhenInUsePermissionScreen',
+          },
+          {
+            shouldShow: shouldShowShareTravelHabitsScreen,
+            screenName: 'Root_ShareTravelHabitsScreen',
+          },
+          {
+            shouldShow: shouldShowNotificationPermissionScreen,
+            screenName: 'Root_NotificationPermissionScreen',
```

Comment:
It can be considered to move `Root_NotificationPermissionScreen` to before `Root_LocationWhenInUsePermissionScreen`. Shouldn't matter too much right now though, and it's a quick fix later on if desired.